### ISSUE TITLE
[Feature] yarn version --deferred should modify the directory of every affected package

### ISSUE THREAD

**Issue by bryanlarsen** [yarnpkg/berry#925]
- I'd be willing to help implement this feature
- This feature would be implemented on the version plugin

**Describe the user story**

Setup:

A workspaces based monorepo.   CI done using GitLab (although the same problem probably exists in other, but not all CI vendors).   test repo:  https://github.com/bryanlarsen/berry

"service" package has dependency "mod".

A developer makes a change to "mod" and then runs `yarn version check --interactive`, resulting in a commit something like https://github.com/bryanlarsen/berry/commit/933e47a664087a5a13be8bc8039081e59c5ef49d

The CI system therefore needs to run the tests for both "mod" and "service".   However, it should not run pipelines for any other packages.

**Describe the solution you'd like**

`yarn version --deferred` when used with workspaces should create a file within each marked package in addition to the files it creates within `.yarn/releases` of the workspaces root.  The CI system can trivially detect that a subdirectory has changed and run appropriate stages.

**Describe the drawbacks of your solution**

- the standard drawbacks of denormalization
- the current version of `yarn version apply` would not clean up the new files created by this hypothetical new version of `yarn version --deferred`

**Describe alternatives you've considered**

- I could write a script stage that parses .yarn/versions and creates a GitLab artifact, but AFAICT surprisingly it seems there is no way for `only:` to trigger off an artifact in GitLab.   Even if there was, this seems like a lot of rigamarole for a seemingly minor problem.

- I could have the CI system push to the repo to do something similar, but having the CI system also creating commits really messes with developers who now have to do a pull before every push

- Instead of having stages not run when they don't need to, I can have them exit early if they don't need to run.   The problem here is that for many packages, starting the stage takes a lot more time than actually running the tests so this doesn't save much.

- I can create a script that wraps `yarn version --deferred` that I force our devs to use.   Probably better just to fork the version plugin instead.   But if I'm going to do this, getting the input from the yarn berry devs is highly useful.   -> this feature request

**Comment by arcanis** [comment]
> `yarn version --deferred` when used with workspaces should create a file within each marked package in addition to the files it creates within `.yarn/releases` of the workspaces root. The CI system can trivially detect that a subdirectory has changed and run appropriate stages.

I don't really want to do this - this is what the previous implementation did, but it caused a huge number of review overhead and file conflicts. The current design with a single file is much easier to work with.

That being said, wouldn't it be possible to have a similar feature to what you're looking for by implementing a `yarn version --json --dry-run --branch-only` command that would return the list of workspaces that would be affected by the current advertised changes? We already have this, except that there's no `--dry-run` and `--branch-only` options.

> I could write a script stage that parses .yarn/versions and creates a GitLab artifact, but AFAICT surprisingly it seems there is no way for `only:` to trigger off an artifact in GitLab.   Even if there was, this seems like a lot of rigamarole for a seemingly minor problem.

Hm I think I understand - you'd like *Gitlab* to trigger one workflow for each workspace to release, is that correct? I'd probably suggest a different workflow where a single workflow is triggered and releases all the relevant packages, if at all possible (that's basically what we do for our own release process). But maybe I didn't understand the problem entirely?

**Comment by bryanlarsen** [comment]
> I don't really want to do this - this is what the previous implementation did, but it caused a huge number of review overhead and file conflicts. The current design with a single file is much easier to work with.

I don't know what you're previous design was like, but AFAICT your current design doesn't use a single file.   It creates multiple files at .yarn/releases/<hash>.yml, which I think is an absolutely brilliant idea.   Each commit gets a new file with a hash filename, so no conflicts.   I'd use the same concept in the subdirs to avoid conflicts if I had to maintain my own fork.

> 
> That being said, wouldn't it be possible to have a similar feature to what you're looking for by implementing a `yarn version --json --dry-run --branch-only` command that would return the list of workspaces that would be affected by the current advertised changes? We already have this, except that there's no `--dry-run` and `--branch-only` options.

I had a long reply written out why this wouldn't work for GitLab CI, but I deleted it because I think I finally figured out how to make this work for GitLab CI.

AFAICT, what we'd have to do is have our monorepo contain a "bootstrap" pipeline.  It'd run your `yarn version --json --dry-run --branch-only` command and then use it to **generate** a new pipeline and tell GitLab to run the new pipeline as a child of the current one.

It'd work but it sounds like a lot of work.   I think it'd be easier for us to maintain our own fork of your version plugin.

> 
> > I could write a script stage that parses .yarn/versions and creates a GitLab artifact, but AFAICT surprisingly it seems there is no way for `only:` to trigger off an artifact in GitLab.   Even if there was, this seems like a lot of rigamarole for a seemingly minor problem.
> 
> Hm I think I understand - you'd like _Gitlab_ to trigger one workflow for each workspace to release, is that correct? I'd probably suggest a different workflow where a single workflow is triggered and releases all the relevant packages, if at all possible (that's basically what we do for our own release process). But maybe I didn't understand the problem entirely?

It's not release so much that's driving this need, it's test.   Some packages have very different and very esoteric needs to run their integration tests:  stuff like X servers, headless browsers, databases, licensed binary-only dependencies, et cetera.   So it's difficult if not impossible to run all of the tests inside the same container.   Not to mention that doing it serially would take hours.

To restate with GitLab terminology: we will have a single "pipeline" in GitLab.   Each "pipeline" in GitLab contains multiple "stages".   (a stage might be a single step or a child pipeline, but that's hopefully irrelevant).   The tests for each package are each a stage in the pipeline that can be run in parallel on different machines.   But GitLab decides at the beginning of the pipeline which stages will need to be run, it can't do that in the middle of the pipeline.

We also have another alternative that I should have listed in my original request:  we could use non-deferred versioning.   That'll bump the `package.json` in every affected dir, just as our CI likes it.   The drawbacks are the reasons that you created deferred versioning in the first place.  :)

**Comment by bryanlarsen** [comment]
Arg, I take some of that back.   The workaround I found to make it work isn't yet available in gitlab.  It's scheduled to be released in the next release, Feb 22.   I hope it doesn't get delayed.   https://gitlab.com/gitlab-org/gitlab/issues/35632
